scripts/pylib/twister/twisterlib/testplan.py

```diff
@@ -618,6 +618,9 @@ def load_from_file(self, file, filter_platform=[]):
                         self.hwm
                     )
 
+                    if self.options.test_only and not instance.run:
```

Comment:
hmm, what is the reason for this ?
it seems that now for `test_only` the resulting report will have these test suites with empty testcases[]  ?

Reply:
seems ok, this is my test result.
```
{
    "environment":{
        "os":"posix",
        "zephyr_version":"v3.7.0-4864-g94df9b68416c",
        "toolchain":"zephyr",
        "commit_date":"2024-10-23T09:00:05+00:00",
        "run_date":"2024-10-23T22:46:13+00:00",
        "options":{
            "device_testing":true,
            "hardware_map":"/home/ubuntu/nxp/frdm_k64f/map.yaml",
            "test_only":true,
            "platform":[
                "frdm_k64f"
            ],
            "west_flash":[]
        }
    },
    "testsuites":[
        {
            "name":"samples/hello_world/sample.basic.helloworld",
            "arch":"arm",
            "platform":"frdm_k64f",
            "path":"samples/hello_world",
            "run_id":"81b66401fdba63c34f93d328a90dbc85",
            "runnable":true,
            "retries":0,
            "dut":"0240000047784e4500229003d917001be561000097969900",
            "status":"passed",
            "execution_time":"4.85",
            "build_time":"0.00",
            "testcases":[
                {
                    "identifier":"sample.basic.helloworld",
                    "execution_time":"4.85",
                    "status":"passed"
                }
            ]
        }
    ]
}

```

Reply:
I think this PR needs rebase to include recent changes done #77250 - looks like there is a regression introduced there for `--test-only` 
I've filed #80332 for that.

Reply:
according to the fix logic it is platform naming issue, which is not related to test status. BTW, what is the full name for qemu_x86?

Reply:
> hmm, what is the reason for this ? it seems that now for `test_only` the resulting report will have these test suites with empty testcases[] ?

This will remove unrunnable test instances from the `--test-only` run. Previously, we've had irrelevant suites reported in the ending summary, which is incorrect.

To illustrate: Imagine that a `--build-only` run contains two test instances. Both are built and achieve `NOTRUN` status. Then, we use Twister in `--test-only` mode. One testcase is runnable and gets a `PASS`. Second one is still not runnable.

Previously, the second instance would be read and included in the report as a `NOTRUN`. This is inconsistent with the `--test-only`'s behaviour when it comes to `SKIP` suites - those are discarded at the start and not included. Now, the unrunnable `NOTRUN` suite will be discarded as well. This is good, because the suite will not be processed either way, so including it in the report is erroneous.

Reply:
> Previously, the second instance would be read and included in the report as a `NOTRUN`. This is inconsistent with the `--test-only`'s behaviour when it comes to `SKIP` suites - those are discarded at the start and not included.
> Now, the unrunnable `NOTRUN` suite will be discarded as well.  This is good, because the suite will not be processed either way, so including it in the report is erroneous.

I'm not sure it is good. In `build-only -> test-only` mode the twister.json is overwritten, so skipped non-runnable will disappear with their built status. Also, with `--detailed-skipped-report` option the 'SKIP'-ed should be included, whereas `NOTRUN`-ed now omitted.

Reply:
Let us compare representative samples of twister run summaries with and without these two lines. I have run Twister with `--build-only` and then `--test-only` flag two times - one with the `if` and once without. I have used a special test set that achieves a more thorough vertical slice of possible test statuses (the same as described in #77488).

Here is the summary without the `if`:
```
INFO    - 36 test scenarios (14 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
--------------------------------
Total test suites: 14
Total test cases: 121
Executed test cases: 120
Skipped test cases: 1
Completed test suites: 13
Passing test suites: 11
Built only test suites: 0
Failing test suites: 2
Skipped test suites: 0
Skipped test suites (runtime): 0
Skipped test suites (filter): 0
Errors: 0
--------------------------------
INFO    - 11 of 14 test configurations passed (78.57%), 0 built (not run), 2 failed, 0 errored, 0 skipped with 0 warnings in 65.25 seconds
INFO    - In total 120 test cases were executed, 1 skipped on 0 out of total 860 platforms (0.00%)
INFO    - 12 test configurations executed on platforms, 2 test configurations were only built.
```


And here with the `if`:
```
INFO    - 36 test scenarios (13 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
--------------------------------
Total test suites: 13
Total test cases: 121
Executed test cases: 120
Skipped test cases: 1
Completed test suites: 13
Passing test suites: 11
Built only test suites: 0
Failing test suites: 2
Skipped test suites: 0
Skipped test suites (runtime): 0
Skipped test suites (filter): 0
Errors: 0
--------------------------------
INFO    - 11 of 13 test configurations passed (84.62%), 0 built (not run), 2 failed, 0 errored, 0 skipped with 0 warnings in 65.40 seconds
INFO    - In total 120 test cases were executed, 1 skipped on 0 out of total 860 platforms (0.00%)
INFO    - 12 test configurations executed on platforms, 1 test configurations were only built.
```

The effect is obvious - we do not consider the unrunnable testsuite that got NOTRUN status during the `--build-only` stage in the `--test-only` stage. It will not be reported on; neither in the console summary, nor in the `twister.json`. This is consistent with the current `--test-only` behaviour when it comes to `SKIP`ped tests. As `--test-only` concerns itself only with runnable tests, as evidenced in the `if` code that resets relevant statuses in `--test-only` mode, this is correct behaviour and should not be removed from this PR.

Without these lines, we'd be pushing still-buggy code to main, which would be unwise.

Reply:
@LukaszMrugala, I don't have your special complex test set to try it out, but just with a simple experiment with one test scenario changed as build_only (`testing.ztest.base.verbose_0` at `tests/ztest/base`) and compare its effect for `twister -T tests/ztest -p qemu_x86_64` run with this PR's code:

with this new `if` condition:
```
build phase
INFO    - 0 test configurations executed on platforms, 10 test configurations were only built.  

test phase:
INFO    - 30 test scenarios (9 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO    - 9 of 9 test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, 0 skipped with 0 warnings in 17.62 seconds
INFO    - 9 test configurations executed on platforms, 0 test configurations were only built.
```
and without it:
```
build phase
INFO    - 0 test configurations executed on platforms, 10 test configurations were only built.  

INFO    - 30 test scenarios (10 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).               
INFO    - 9 of 10 test configurations passed (90.00%), 0 built (not run), 0 failed, 0 errored, 0 skipped with 0 warnings in 20.43 seconds                                                                                                                                 
INFO    - 9 test configurations executed on platforms, 1 test configurations were only built.                                        
```
please note:
1) both show 10 after build
2) the latter shows, that the built only test configuration (1 of 10) is present in the final report, whereas in the former case it disappears (only 9 on 10) and it is wrong.

There is still a common problem though: both summaries have `0 built (not run)` instead of `1 built (not run)`.

> The effect is obvious - we do not consider the unrunnable testsuite that got NOTRUN status during the --build-only stage in the --test-only stage.  It will not be reported on; neither in the console summary, nor in the twister.json.

all suites are NOTRUN after `--build-only` and I don't think it is correct to hide built-only instances not reporting them after `--test-only` phase done on all the artifacts actually built.